lib/change_stream_simulator.js

```diff
@@ -0,0 +1,244 @@
+var Timestamp = require('mongodb-core').BSON.Timestamp;
+var MongoNetworkError = require('mongodb-core').MongoNetworkError;
+var supportedStages =  ['$match', '$project', '$addFields', '$replaceRoot', '$redact'];
+
+var createChangeStreamCursorSimulation = function (serverConfig, namespace, pipeline, options) {
+  var DB = require('./db');
+
+  // Validate inputs
+  if (typeof serverConfig !== 'object') throw new Error('A server topology object must be provided to createChangeStreamCursorSimulation');
+  if (!(typeof namespace === 'object' && typeof namespace.database === 'string')) throw new Error('A namespace object containing a database name must be provided to createChangeStreamCursorSimulation');
+
+  // Verify the server topology is a replica set
+  if (!serverConfig.s.replset) {
+    throw new Error('Change Stream are only supported on replica sets. The connected server does not appear to be a replica set.');
+  }
+
+  // Verify all the pipeline stages are supported
+  pipeline.forEach(function (stage) {
+    var stageName = Object.keys(stage)[0];
+    if (supportedStages.indexOf(stageName) === -1) {
+      throw new Error('The pipeline contains the stage "' + stageName + '", which is not compatible with Change Streams at this time.');
+    }
+  });
+
+  // Connect to the oplog
+  var localDB = new DB('local', serverConfig);
+  var oplogCollection = localDB.collection('oplog.rs');
+  var aggregationProcessorCollection = localDB.collection('aggregationProcessor' + options.cursorNoConflict);
+  var aggregationOutputCollection = localDB.collection('aggregationOutput' + options.cursorNoConflict);
+
+  // Ensure the existence of the aggregationOutput capped collection with at least one item
+  localDB.createCollection('aggregationOutput' + options.cursorNoConflict, { capped: true, size: 1000, max: 10 }, function (err, collection) {
+    if (collection) {
+      collection.insert({seed:true});
+    }
+  });
+
+  // Construct a basic filter of the oplog
```

Comment:
it kind of looks like a lot of this function could be broken apart into various smaller self-documenting functions. For instance, it looks like this section is specific to creating the tailable cursor for the oplog